stdlib/public/core/TypeLayoutConstraints.swift

```diff
@@ -0,0 +1,40 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+// Define some compile-known protocols used to constraint type layout.
+// These protocols can be used e.g. in @_specialize attribute.
+
+// Type should be a trivial type. It's size should be 8 bits.
+public protocol _Trivial8 {
```

Comment:
I see your other comment now. Yeah, we should find a way to represent these constraints as their own thing and not protocols. You already added code to ArchetypeBuilder to handle them -- why are the protocols needed?